platforms/jvm/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java

```diff
@@ -659,6 +662,10 @@ public ModularitySpec getModularity() {
      */
     @Override
     protected JvmTestExecutionSpec createTestExecutionSpec() {
+        if (!getTestFramework().supportsNonClassBasedTesting() && getScanForTestDefinitions().get()) {
```

Comment:
I think this should also fail if `getCandidateDefinitionDirs()` or `getTestDefinitionDirs()` has anything in it.

Reply:
Can we then add tests for JUnit 4 and TestNG to verify that if dirs are provided but no (test) classes, it fails with the "no tests run" message?